File: scripts/sync-traders-to-bigquery.js
#!/usr/bin/env node

/**
 * Sync all wallet addresses from Supabase traders table to BigQuery traders table
 */

require('dotenv').config({ path: '.env.local' })
const { createClient } = require('@supabase/supabase-js')
const { execSync } = require('child_process')
const fs = require('fs')

const SUPABASE_URL = process.env.NEXT_PUBLIC_SUPABASE_URL
const SUPABASE_SERVICE_ROLE_KEY = process.env.SUPABASE_SERVICE_ROLE_KEY
const PROJECT_ID = 'gen-lang-client-0299056258'
const DATASET = 'polycopy_v1'
const TABLE = 'traders'

if (!SUPABASE_URL || !SUPABASE_SERVICE_ROLE_KEY) {
  console.error('❌ Missing NEXT_PUBLIC_SUPABASE_URL or SUPABASE_SERVICE_ROLE_KEY')
  process.exit(1)
}

const supabase = createClient(SUPABASE_URL, SUPABASE_SERVICE_ROLE_KEY)

async function main() {
  console.log('='.repeat(60))
  console.log('🔄 Syncing traders from Supabase to BigQuery')
  console.log('='.repeat(60))
  
  try {
    // Fetch all wallet addresses from Supabase traders table
    console.log('\n📊 Fetching wallet addresses from Supabase...')
    const { data: traders, error } = await supabase
      .from('traders')
      .select('wallet_address')
      .not('wallet_address', 'is', null)
    
    if (error) {
      console.error('❌ Error fetching traders:', error.message)
      process.exit(1)
    }
    
    if (!traders || traders.length === 0) {
      console.log('⚠️  No traders found in Supabase')
      return
    }
    
    // Extract unique wallet addresses
    const wallets = [...new Set(traders.map(t => t.wallet_address?.toLowerCase().trim()).filter(Boolean))]
    console.log(`✅ Found ${wallets.length} unique wallet addresses`)
    
    if (wallets.length === 0) {
      console.log('⚠️  No valid wallet addresses found')
      return
    }
    
    // Check existing wallets in BigQuery
    console.log('\n📊 Checking existing wallets in BigQuery...')
    const checkQuery = `SELECT wallet_address FROM \\\`${PROJECT_ID}.${DATASET}.${TABLE}\\\``
    let existingWallets = new Set()
    
    try {
      const checkResult = execSync(
        `bq query --use_legacy_sql=false --format=json --project_id=${PROJECT_ID} '${checkQuery}'`,
        { encoding: 'utf-8' }
      )
      const existing = JSON.parse(checkResult)
      if (existing && existing.length > 0) {
        existingWallets = new Set(existing.map(row => row.wallet_address?.toLowerCase()))
      }
    } catch (e) {
      // Table might not exist or be empty, that's okay
      console.log('ℹ️  No existing wallets found (or table is empty)')
    }
    
    // Filter out wallets that already exist
    const newWallets = wallets.filter(w => !existingWallets.has(w.toLowerCase()))
    console.log(`📊 New wallets to insert: ${newWallets.length}`)
    console.log(`📊 Already in BigQuery: ${wallets.length - newWallets.length}`)
    
    if (newWallets.length === 0) {
      console.log('\n✅ All wallets already synced to BigQuery!')
      return
    }
    
    // Insert in batches of 1000 (BigQuery limit)
    const BATCH_SIZE = 1000
    let inserted = 0
    
    // Write wallets to JSON file for bq load
    const fs = require('fs')
    const tempFile = '/tmp/traders_wallets.json'
    
    for (let i = 0; i < newWallets.length; i += BATCH_SIZE) {
      const batch = newWallets.slice(i, i + BATCH_SIZE)
      // NEWLINE_DELIMITED_JSON requires one JSON object per line
      const jsonLines = batch.map(w => JSON.stringify({ wallet_address: w })).join('\n')
      
      // Write batch to JSON file (one object per line)
      fs.writeFileSync(tempFile, jsonLines)
      
      console.log(`\n📤 Inserting batch ${Math.floor(i / BATCH_SIZE) + 1} (${batch.length} wallets)...`)
      
      try {
        // Use bq load to insert from JSON file
        execSync(
          `bq load --source_format=NEWLINE_DELIMITED_JSON --project_id=${PROJECT_ID} --replace=false ${PROJECT_ID}:${DATASET}.${TABLE} ${tempFile} wallet_address:STRING`,
          { encoding: 'utf-8', stdio: 'inherit' }
        )
        inserted += batch.length
        console.log(`✅ Inserted ${batch.length} wallets (${inserted}/${newWallets.length} total)`)
      } catch (e) {
        console.error(`❌ Error inserting batch:`, e.message)
        throw e
      }
    }
    
    // Clean up temp file
    try {
      fs.unlinkSync(tempFile)
    } catch (e) {
      // Ignore cleanup errors
    }
    
    console.log('\n' + '='.repeat(60))
    console.log(`✅ Successfully synced ${inserted} wallet addresses to BigQuery!`)
    console.log(`📊 Total wallets in BigQuery: ${wallets.length}`)
    console.log('='.repeat(60))
    
  } catch (error) {
    console.error('\n❌ Error:', error.message)
    process.exit(1)
  }
}

main()

File: scripts/sync-traders-to-bigquery.py
#!/usr/bin/env python3

"""
Sync all wallet addresses from Supabase traders table to BigQuery traders table
"""

import os
import sys
from supabase import create_client, Client
from google.cloud import bigquery
from dotenv import load_dotenv

load_dotenv('.env.local')

SUPABASE_URL = os.getenv('NEXT_PUBLIC_SUPABASE_URL')
SUPABASE_SERVICE_ROLE_KEY = os.getenv('SUPABASE_SERVICE_ROLE_KEY')
PROJECT_ID = 'gen-lang-client-0299056258'
DATASET = 'polycopy_v1'
TABLE = 'traders'

if not SUPABASE_URL or not SUPABASE_SERVICE_ROLE_KEY:
    print('❌ Missing NEXT_PUBLIC_SUPABASE_URL or SUPABASE_SERVICE_ROLE_KEY')
    sys.exit(1)

supabase: Client = create_client(SUPABASE_URL, SUPABASE_SERVICE_ROLE_KEY)
bq_client = bigquery.Client(project=PROJECT_ID)

def main():
    print('=' * 60)
    print('🔄 Syncing traders from Supabase to BigQuery')
    print('=' * 60)
    
    # Fetch all wallet addresses from Supabase (handle pagination)
    print('\n📊 Fetching wallet addresses from Supabase...')
    response = supabase.table('traders').select('wallet_address', count='exact').not_.is_('wallet_address', 'null').execute()
    
    # Get all pages if needed
    traders = []
    total_count = response.count if response.count else len(response.data or [])
    
    if total_count > 1000:
        # Fetch in batches
        print(f"  📊 Fetching {total_count} traders in batches...")
        for offset in range(0, total_count, 1000):
            batch = supabase.table('traders').select('wallet_address').not_.is_('wallet_address', 'null').range(offset, min(offset + 999, total_count - 1)).execute()
            traders.extend(batch.data or [])
            if (offset // 1000 + 1) % 5 == 0:
                print(f"     Fetched {len(traders)}/{total_count}...")
    else:
        traders = response.data or []
    
    if not traders:
        print('⚠️  No traders found in Supabase')
        return
    
    # Extract unique wallet addresses
    wallets = list(set([t['wallet_address'].lower().strip() for t in traders if t.get('wallet_address')]))
    print(f'✅ Found {len(wallets)} unique wallet addresses')
    
    if not wallets:
        print('⚠️  No valid wallet addresses found')
        return
    
    # Check existing wallets in BigQuery
    print('\n📊 Checking existing wallets in BigQuery...')
    table_id = f'{PROJECT_ID}.{DATASET}.{TABLE}'
    existing_wallets = set()
    
    try:
        query = f"SELECT DISTINCT wallet_address FROM `{table_id}`"
        results = bq_client.query(query).result()
        existing_wallets = {row.wallet_address.lower() for row in results if row.wallet_address}
    except Exception as e:
        print(f'ℹ️  No existing wallets found (or table is empty): {e}')
    
    # Filter out wallets that already exist
    new_wallets = [w for w in wallets if w.lower() not in existing_wallets]
    print(f'📊 New wallets to insert: {len(new_wallets)}')
    print(f'📊 Already in BigQuery: {len(wallets) - len(new_wallets)}')
    
    if not new_wallets:
        print('\n✅ All wallets already synced to BigQuery!')
        return
    
    # Insert in batches
    BATCH_SIZE = 1000
    inserted = 0
    
    for i in range(0, len(new_wallets), BATCH_SIZE):
        batch = new_wallets[i:i + BATCH_SIZE]
        rows_to_insert = [{'wallet_address': w} for w in batch]
        
        print(f'\n📤 Inserting batch {i // BATCH_SIZE + 1} ({len(batch)} wallets)...')
        
        try:
            errors = bq_client.insert_rows_json(table_id, rows_to_insert)
            if errors:
                print(f'❌ Errors inserting batch: {errors}')
                raise Exception(f'Insert errors: {errors}')
            
            inserted += len(batch)
            print(f'✅ Inserted {len(batch)} wallets ({inserted}/{len(new_wallets)} total)')
        except Exception as e:
            print(f'❌ Error inserting batch: {e}')
            raise
    
    print('\n' + '=' * 60)
    print(f'✅ Successfully synced {inserted} wallet addresses to BigQuery!')
    print(f'📊 Total wallets in BigQuery: {len(wallets)}')
    print('=' * 60)

if __name__ == '__main__':
    main()
